File: ruby/matrix/matrix.rb
class Matrix
  def initialize(matrix)
    @matrix = matrix
  end

  def rows
    @rows ||= from_matrix_representation
  end

  def columns
    @columns ||= rows.transpose
  end

  private

  attr_reader :matrix

  def from_matrix_representation
    matrix.each_line.map do |line|
      line.split.map(&:to_i)
    end
  end
end

File: javascript/matrix/matrix.js
export class Matrix {
  constructor (str) {
    this.rows = strToMatrixOfNumbers(str)
    this.columns = transpose(this.rows)
  }
}

function strToMatrixOfNumbers (str) {
  return str.split('\n').map(line => {
    return line.split(' ').map(value => Number.parseInt(value, 10))
  })
}

function transpose (matrix) {
  const transpose = []
  for (let i = 0; i < matrix.length; i++) {
    transpose.push([])
  }

  for (let line of matrix) {
    for (let i = 0; i < line.length; i++) {
      transpose[i].push(line[i])
    }
  }
  return transpose
}
